### ISSUE TITLE
PainlessScript$Script has illegal signature "(Ljava/lang/Object;V)V crashes elasticsearch

### ISSUE THREAD

**Issue by jpcarey** [elastic/elasticsearch#27210]
**Elasticsearch version**: elasticsearch-5.6.3
**Plugins installed**: x-pack
**JVM version**:
```
java version "1.8.0_65"
Java(TM) SE Runtime Environment (build 1.8.0_65-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.65-b01, mixed mode)
```

**OS version**:
`Darwin Kernel Version 16.7.0; root:xnu-3789.70.16~2/RELEASE_X86_64 x86_64`

**behavior**: Crash / exit elasticsearch:

```
[2017-11-01T09:26:30,732][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [] fatal error in thread [elasticsearch[1g9MtQh][management][T#2]], exiting
java.lang.ClassFormatError: Method "start" in class org/elasticsearch/painless/PainlessScript$Script has illegal signature "(Ljava/lang/Object;V)V"
	at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.8.0_65]
	at java.lang.ClassLoader.defineClass(ClassLoader.java:760) ~[?:1.8.0_65]
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142) ~[?:1.8.0_65]
	at org.elasticsearch.painless.Compiler$Loader.defineScript(Compiler.java:86) ~[?:?]
	at org.elasticsearch.painless.Compiler.compile(Compiler.java:134) ~[?:?]
	at org.elasticsearch.painless.PainlessScriptEngineService$2.run(PainlessScriptEngineService.java:176) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_65]
	at org.elasticsearch.painless.PainlessScriptEngineService.compile(PainlessScriptEngineService.java:172) ~[?:?]
	at org.elasticsearch.painless.PainlessScriptEngineService.compile(PainlessScriptEngineService.java:112) ~[?:?]
	at org.elasticsearch.script.ScriptService.compile(ScriptService.java:307) ~[elasticsearch-5.6.3.jar:5.6.3]
	at org.elasticsearch.xpack.watcher.condition.ScriptCondition.<init>(ScriptCondition.java:59) ~[?:?]
	at org.elasticsearch.xpack.watcher.condition.ScriptCondition.parse(ScriptCondition.java:75) ~[?:?]
	at org.elasticsearch.xpack.watcher.Watcher.lambda$createComponents$4(Watcher.java:229) ~[?:?]
	at org.elasticsearch.xpack.watcher.condition.ConditionRegistry.parseExecutable(ConditionRegistry.java:74) ~[?:?]
	at org.elasticsearch.xpack.watcher.watch.Watch$Parser.parse(Watch.java:316) ~[?:?]
	at org.elasticsearch.xpack.watcher.watch.Watch$Parser.parse(Watch.java:282) ~[?:?]
	at org.elasticsearch.xpack.watcher.watch.Watch$Parser.parse(Watch.java:241) ~[?:?]
	at org.elasticsearch.xpack.watcher.transport.actions.execute.TransportExecuteWatchAction.masterOperation(TransportExecuteWatchAction.java:118) ~[?:?]
	at org.elasticsearch.xpack.watcher.transport.actions.execute.TransportExecuteWatchAction.masterOperation(TransportExecuteWatchAction.java:65) ~[?:?]
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.masterOperation(TransportMasterNodeAction.java:87) ~[elasticsearch-5.6.3.jar:5.6.3]
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.doRun(TransportMasterNodeAction.java:166) ~[elasticsearch-5.6.3.jar:5.6.3]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) ~[elasticsearch-5.6.3.jar:5.6.3]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-5.6.3.jar:5.6.3]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_65]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_65]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_65]
```
**Steps to reproduce**:

``` json
POST crash-test/doc/_bulk
{"index": {}}
{ "@timestamp" : "2017-10-20T18:26:41.431Z", "type": "nginx" }
{"index": {}}
{ "@timestamp" : "2017-10-20T17:55:29.190Z", "type": "nginx" }
```

```
POST _xpack/watcher/watch/_execute
{
  "watch": {
    "metadata": {
      "indexing_threshold": 20
    },
    "trigger": {
      "schedule": { "daily": { "at": "2:00" }}
    },
    "input": {
      "search": {
        "request": {
          "indices": [ "crash-test" ],
          "body": {
            "size": 0,
            "aggs": {
              "TYPE": {
                "terms": {
                  "field": "type.keyword",
                  "size": 10
                },
                "aggs": {
                  "MAX": {
                    "max": {
                      "field": "@timestamp"
                    }
                  }
                }
              }
            }
          }
        }
      }
    },
    "condition": {
      "script": {
        "source": """
ctx.vars.bad = ctx.payload.aggregations.TYPE.buckets.stream()
  .map( it -> {
    Calendar calendar = Calendar.getInstance();
    calendar.setTimeInMillis((long) it.MAX.value);
    it.start = calendar.add(Calendar.MINUTE, 5);
    return it
  })
  .collect(Collectors.toList());

return true;
"""
      }
    },
    "actions": {
      "test_output": {
        "logging": {
          "text": "{{ ctx.vars.bad }}"
        }
      }
    }
  }
}
```

**Comment by jdconrad** [comment]
So, it looks like the add method of Calendar returns void which means ```it``` cannot be assigned to at the line ```it.start = calendar.add(Calendar.MINUTE, 5);```.  I would consider this a bug in Painless because this error message is probably the *worst* one I've seen so far, and the error should never get as far compilation in this case.  I'll need to investigate further, but as a guess this happens because it is of the def type, and we assume any type can be assigned to a def when a void actually cannot be under any circumstance.

**Comment by jdconrad** [comment]
This is fixed and backported all the way back to 5.6.5.

**Comment by jasontedor** [comment]
Thanks @jdconrad.
